test/framework/toy_build.py

```diff
@@ -2025,6 +2027,140 @@ def test_toy_build_hooks(self):
+            self.modtool.unload(['toy/0.0', 'GCC/4.6.3'])
+
+            # if GCC/7.3.0-2.30 is loaded first, then GCC/4.6.3 is not loaded by loading toy/0.0
+            self.modtool.load(['GCC/7.3.0-2.30'])
+            loaded_mod_names = [x['mod_name'] for x in self.modtool.list()]
+            self.assertTrue('GCC/7.3.0-2.30' in loaded_mod_names)
+
+            self.modtool.load(['toy/0.0'])
+            loaded_mod_names = [x['mod_name'] for x in self.modtool.list()]
+            self.assertTrue('toy/0.0' in loaded_mod_names)
+            self.assertTrue('GCC/7.3.0-2.30' in loaded_mod_names)
+            self.assertFalse('GCC/4.6.3' in loaded_mod_names)
+
+            self.modtool.unload(['toy/0.0', 'GCC/7.3.0-2.30'])
+
+            # of course having GCC/4.6.3 loaded already is also fine
+            self.modtool.load(['GCC/4.6.3'])
+            loaded_mod_names = [x['mod_name'] for x in self.modtool.list()]
+            self.assertTrue('GCC/4.6.3' in loaded_mod_names)
+
+            self.modtool.load(['toy/0.0'])
+            loaded_mod_names = [x['mod_name'] for x in self.modtool.list()]
+            self.assertTrue('toy/0.0' in loaded_mod_names)
+            self.assertTrue('GCC/4.6.3' in loaded_mod_names)
+            self.assertFalse('GCC/7.3.0-2.30' in loaded_mod_names)
+
+            self.modtool.unload(['toy/0.0', 'GCC/4.6.3'])
+
+        check_toy_load()
+
+        # this behaviour can be disabled via "multi_dep_load_defaults = False"
+        write_file(test_ec, test_ec_txt + "\nmulti_deps_load_default = False")
+
+        remove_file(toy_mod_file)
+        self.test_toy_build(ec_file=test_ec)
+        toy_mod_txt = read_file(toy_mod_file)
+
+        self.assertFalse(expected in toy_mod_txt, "Pattern '%s' should not be found in: %s" % (expected, toy_mod_txt))
+
+        self.modtool.load(['toy/0.0'])
+        loaded_mod_names = [x['mod_name'] for x in self.modtool.list()]
+        self.assertTrue('toy/0.0' in loaded_mod_names)
+        self.assertFalse('GCC/4.6.3' in loaded_mod_names)
+        self.assertFalse('GCC/7.3.0-2.30' in loaded_mod_names)
+
+        # restore original environment to continue testing with a clean slate
+        modify_env(os.environ, self.orig_environ, verbose=False)
+        self.modtool.use(test_mod_path)
+
+        write_file(test_ec, test_ec_txt)
+
+        # also check behaviour when using 'depends_on' rather than 'load' statements (requires Lmod 7.6.1 or newer)
+        if self.modtool.supports_depends_on:
+
+            remove_file(toy_mod_file)
+            self.test_toy_build(ec_file=test_ec, extra_args=['--module-depends-on'])
+
+            toy_mod_txt = read_file(toy_mod_file)
+
+            # check whether (guarded) load statement for first version listed in multi_deps is there
+            if get_module_syntax() == 'Lua':
+                cond = 'mode() == "unload" or ( not ( isloaded("GCC/4.6.3") ) and not ( isloaded("GCC/7.3.0-2.30") ) )'
```

Comment:
I am not sure I follow the lmod depends_on logic, what happens if one of the two modules is already loaded, then you get no depends_on on load but a depends_on on unload.

And what happens if the user switches from one to another GCC version via module load.

I'll need to try that myself.. but will do so tomorrow morning.

Reply:
Good point...

When the module is being loaded, and no `GCC` is loaded yet, then the first listed version should be loaded as default; that bit is OK with the current approach, both with `load` and `depends_on`.

When the module is being unloaded, the `GCC` that is loaded should be unloaded; that bit is not OK with `depends_on` right now, since we will blindly lower the reference count for the "default" version, not for the one that is loaded...

So, on `unload` we just need a `depends_on("GCC")`? Would that work?

We do *not* want to use that on `load`, since we want to control which version gets loaded (it may not be the latest version/default-in-the-Lmod-sense).

So, something like this:

```lua
if not ( isloaded("GCC/4.6.3") ) and not ( isloaded("GCC/7.3.0-2.30") ) then
    depends_on("GCC/4.6.3")
elseif mode() == "unload" then
    depends_on("GCC")
end
```

Reply:
Actually, it can be done a lot cleaner:

```lua
if isloaded("Python/2.7.15-GCCcore-8.2.0") then
    depends_on("Python/2.7.15-GCCcore-8.2.0")
else
    depends_on("Python/3.7.2-GCCcore-8.2.0")
end
```

This way, we bump the ref count for `Python/2.7.15-GCCcore-8.2.0` if it's loaded already. Same for `Python/3.7.2-GCCcore-8.2.0`, but since there's no guard there, it'll also get loaded as a default if neither are loaded.

On unloading, it also works as expected: it will lower the ref count for `Python/2.7.15-GCCcore-8.2.0` if that one is loaded, or lower the ref count (perhaps also unloading) `Python/3.7.2-GCCcore-8.2.0` otherwise.

Reply:
Yes I was thinking on something like that too (there needs to be a `depends_on` for all `if` branches for sure), but I think there may still be an issue if the user switches the python version after loading scipy-bundle; a `depends_on("Python")` for `unload` as you did before may fix this. To be tested...
```
if mode() == "unload" then
    depends_on("Python")
elseif isloaded("Python/2.7.15-GCCcore-8.2.0") then
    depends_on("Python/2.7.15-GCCcore-8.2.0")
else
    depends_on("Python/3.7.2-GCCcore-8.2.0")
end
```

Reply:
@bartoldeman I think my last comment that just uses `isloaded("GCC")` also covers that...

Just pushed the necessary changes with c6387cd31, please re-review?

Reply:
no, now it will behave differently if an incompatible python version is already loaded, and keep using that.

Reply:
This is the shortest I can make it:
```
if mode() == "unload" or isloaded("Python/2.7.15-GCCcore-8.2.0") then
    depends_on("Python")
else
    depends_on("Python/3.7.2-GCCcore-8.2.0")
end
```
List any other non-default version in the or statement, eg. `if mode() == "unload" or isloaded("Python/2.7.15-GCCcore-8.2.0") or isloaded("Python/3.6.6-GCCcore-8.2.0") then`